saw-core-coq/src/Verifier/SAW/Translation/Coq/Term.hs

```diff
@@ -530,92 +579,94 @@ translateImplicitHyp ::
-     names <- traverse (const nextName) locals
-     modify $ set sharedNames names
-     defs <- traverse translateTermUnshared locals
+  let occ_map = scTermCount False t
+      shares = IntMap.assocs $ fmap fst $ IntMap.filter keep occ_map
+      share_tms = map snd shares in
+  -- NOTE: larger terms always have later stAppIndices than their subterms, so
+  -- IntMap.elems above is guaranteed to return subterms before superterms; this
+  -- ensures that the right-hand sides in our nested let-bindings below only
+  -- refer to variables bound earlier, not later
+  withSharedTerms shares $ \names ->
+  do defs <- traverse translateTermUnshared share_tms
      body <- translateTerm t
-     -- NOTE: Larger terms always have later IDs than their subterms,
-     -- so ordering by VarIndex is a valid dependency order.
-     let binds = IntMap.elems (IntMap.intersectionWith (,) names defs)
-     pure (foldr mkLet body binds)
+     pure (foldr mkLet body $ zip names defs)
   where
     keep (t', n) = n > 1 && shouldMemoizeTerm t'
-    nextName =
-      do x <- view nextSharedName <$> get
-         x' <- freshVariant x
-         modify $ set nextSharedName (nextVariant x')
-         pure x'
 
+-- | Translate a SAW core 'Term' to Coq, using let-bound Coq names when they are
+-- associated with the given term for sharing
 translateTerm :: TermTranslationMonad m => Term -> m Coq.Term
 translateTerm t =
   case t of
     Unshared {} -> translateTermUnshared t
     STApp { stAppIndex = i } ->
-      do shared <- view sharedNames <$> get
+      do shared <- view sharedNames <$> askTrr
          case IntMap.lookup i shared of
            Nothing -> translateTermUnshared t
-           Just x -> pure (Coq.Var x)
+           Just sh -> pure (Coq.Var $ sharedNameIdent sh)
 
+-- | Translate a SAW core 'Term' to Coq without using sharing
 translateTermUnshared :: TermTranslationMonad m => Term -> m Coq.Term
-translateTermUnshared t = withLocalTranslationState $ do
+translateTermUnshared t = do
   -- traceTerm "translateTerm" t $
   -- NOTE: env is in innermost-first order
-  env <- view localEnvironment <$> get
+  env <- view localEnvironment <$> askTrr
   -- let t' = trace ("translateTerm: " ++ "env = " ++ show env ++ ", t =" ++ showTerm t) t
   -- case t' of
   case unwrapTermF t of
 
     FTermF ftf -> flatTermFToExpr ftf
 
-    Pi {} -> translatePi params e
-      where
-        (params, e) = asPiList t
+    Pi {} ->
+      let (params, e) = asPiList t in
+      translatePi params e
 
-    Lambda {} -> do
-      paramTerms <- translateParams params
-      e' <- translateTermLet e
-      pure (Coq.Lambda paramTerms e')
-        where
-          -- params are in normal, outermost first, order
-          (params, e) = asLambdaList t
+    Lambda {} ->
+      let (params, e) = asLambdaList t in
+      translateParams params $ \paramTerms ->
+      do e' <- translateTermLet e
+         return (Coq.Lambda paramTerms e')
```

Comment:
Why not put `translateParams` inside the do-block? Or have the whole thing be a do-block including the `let`?

Reply:
Because the body of `translateParams` is a separate computation, that runs in the context of the additional parameters being bound to SAW variables. Does that make sense?

Reply:
Oh right, I was reading this wrong. I got confused because the later `do` was not indented further in, so my brain tricked me. Consider indenting, unless you dislike it?

Reply:
Ok, that makes sense. Mostly I just follow whatever indentation emacs does, but I see what you mean, that that would make this easier to read.

Reply:
Done!